pkg/controller/controller.go

```diff
@@ -352,13 +364,104 @@ func (s *Service) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest
 				return nil, status.Errorf(codes.Internal, "can't query remote system by name: %s", err.Error())
 			}
 
-			// TODO If volumeGroup input is specified in SC - Verify VolumeGroup exists, if not create one
-			// There shouldn't be any protection policy on it with replication rule
-			// Cannot configure Metro on empty VG. Configure after volume is added
-			// Check if the above sync/async block can be optimized w.r.t volume group calls
-			// isMetroVolumeGroup = true
+			// Metro volume group is desired if volumeGroupPrefix storage-class param is set.
+			vgPrefix, ok := params[s.WithRP(KeyReplicationVGPrefix)]
+			if !ok || vgPrefix == "" {
+				isMetroVolume = true
+				startMetroReplicationSession = configureMetroVolumeSession
+			} else {
+				isMetroVolumeGroup = true
+
+				// Add the PVC namespace to the volume group name if storage class param ignoreNamespaces is "false"
+				namespace := ""
+				if ignoreNS, ok := params[s.WithRP(KeyReplicationIgnoreNamespaces)]; ok && ignoreNS == "false" {
+					pvcNS, ok := params[KeyCSIPVCNamespace]
+					if ok {
+						namespace = pvcNS + "-"
+					}
+				}
+
+				// Get the volume group name used for querying PowerStore array
+				vgName := vgPrefix + "-" + namespace + remoteSystemName
+				// truncate name length to comply with CSI gRPC spec
+				if len(vgName) > 128 {
+					vgName = vgName[:128]
+					log.Info("volume group name was truncated because it exceeded the max allowed length.")
+				}
+
+				// check if vg exists on powerstore array
+				vg, err = arr.Client.GetVolumeGroupByName(ctx, vgName)
+				if err == nil {
+					// A volume group with the vgName already exists
+					// Having a protection policy associated with the vg makes it invalid for metro replication.
+					if vg.ProtectionPolicyID != "" {
+						return nil, status.Errorf(codes.FailedPrecondition,
+							"volume group %s has a protection policy assigned making it incompatible for usage with metro replication.", vg.Name)
+					}
+
+					// Confirm VG is part of a metro session and pause the session to allow adding a volume to the vg
+					if vg.MetroReplicationSessionID != "" {
+						// make sure session is in good state
+						metroVGSession, err := arr.Client.GetReplicationSessionByID(ctx, vg.MetroReplicationSessionID)
+						if metroVGSession.State != gopowerstore.RsStateOk {
```

Comment:
Can we handle the case if the VG is already in paused state, to support idempotency?
The group may be paused and then if there is a failure during add volume - the controller will repeat the create volume call.